narwhal/primary/src/synchronizer.rs

```diff
@@ -826,3 +859,163 @@ impl Synchronizer {
+        if let Some(suspended_cert) = self.suspended.remove(&digest) {
+            panic!(
+                "Suspended certificate {digest:?} has no missing parent ({:?} exist in store)",
+                suspended_cert.missing_parents
+            )
+        }
+        let mut to_traverse = VecDeque::new();
+        let mut to_accept = Vec::new();
+        to_traverse.push_back((round, digest));
+        while let Some((round, digest)) = to_traverse.pop_front() {
+            let Some(child_digests) = self.missing.remove(&(round, digest)) else {
+                continue;
+            };
+            for child in &child_digests {
+                let Some(suspended_cert) = self.suspended.get_mut(child) else {
+                    continue;
+                };
+                suspended_cert.missing_parents.remove(&digest);
+                if suspended_cert.missing_parents.is_empty() {
+                    let suspended_cert = self.suspended.remove(child).unwrap();
+                    to_traverse.push_back((
+                        suspended_cert.certificate.round(),
+                        suspended_cert.certificate.digest(),
+                    ));
+                    to_accept.push(suspended_cert);
+                }
+            }
+        }
+        to_accept
+    }
+
+    /// Runs GC on the suspended certificates, returns a list that can be accepted at gc round + 1.
+    /// It is caller's responsibility to check if some children of the returned certificates can
+    /// also be accepted.
+    fn run_gc(&mut self, gc_round: Round) -> Vec<SuspendedCertificate> {
+        // Remove suspended certificates below gc round, and collect digests for certificates just
+        // above the gc round.
+        let mut gc_certificates = Vec::new();
+        let mut certificates_above_gc_round = HashSet::new();
+        while let Some(((round, digest), children)) = self.missing.iter().next() {
+            if *round > gc_round {
+                break;
+            }
+            if *round == gc_round {
+                certificates_above_gc_round.extend(children.iter().cloned());
+            }
+            // It is ok to notify waiters here (via Drop). The certificate will never and does
+            // not need to get into certificate store.
+            if let Some(suspended) = self.suspended.remove(digest) {
+                gc_certificates.push(suspended);
+            }
+            self.missing.remove(&(*round, *digest));
+        }
+        // Notify waiters on GC'ed certificates.
+        for suspended in gc_certificates {
+            let _ = suspended.notify.send(());
+        }
+        // All certificates at gc round + 1 can be accepted.
+        let mut to_accept = Vec::new();
+        for digest in certificates_above_gc_round {
+            let Some(mut suspended_cert) = self.suspended.remove(&digest) else {
+                error!("Suspended certificate {digest} not found!");
+                continue;
+            };
```

Comment:
How is this possible?

Reply:
Right, this should be impossible. Changing to `expect()`.